src/utils/arrow.rs

```diff
@@ -1,105 +1,138 @@
+use std::collections::HashMap;
+use std::sync::Mutex;
+
 use arrow2::compute::cast;
 
-pub fn cast_array_if_needed(
-    arrow_array: Box<dyn arrow2::array::Array>,
-) -> Box<dyn arrow2::array::Array> {
-    match arrow_array.data_type() {
-        arrow2::datatypes::DataType::Utf8 => {
-            cast::utf8_to_large_utf8(arrow_array.as_any().downcast_ref().unwrap()).boxed()
-        }
-        arrow2::datatypes::DataType::Binary => cast::binary_to_large_binary(
-            arrow_array.as_any().downcast_ref().unwrap(),
-            arrow2::datatypes::DataType::LargeBinary,
-        )
-        .boxed(),
+// TODO(Clark): Refactor to GILOnceCell in order to avoid deadlock between the below mutex and the Python GIL.
+lazy_static! {
+    static ref REGISTRY: Mutex<HashMap<std::string::String, arrow2::datatypes::DataType>> =
+        Mutex::new(HashMap::new());
```

Comment:
Hmm I can see how this could lead to a deadlock with the GIL. Could be worthwhile to Fix with the GILOnceCell.